glean-core/python/glean/_loader.py

```diff
@@ -90,6 +90,44 @@ def __getattr__(self, attr):
         )
 
 
+def event_extra_factory(name, argnames):
+    """
+    Generate a new class, inheriting from `metrics.EventExtraClass`
+    and implementing the `to_ffi_extra` method,
+    which serializes expected attributes to pass over FFI.
+    """
+
+    def __init__(self, **kwargs):
+        for key, value in kwargs.items():
+            if key not in argnames:
+                raise TypeError(
+                    "Argument %s not valid for %s" % (key, self.__class__.__name__)
+                )
+            setattr(self, key, value)
+
+    def to_ffi_extra(self):
+        keys = []
+        values = []
+
+        for idx, name in enumerate(argnames):
+            attr = getattr(self, name, None)
+            if attr is not None:
+                keys.append(idx)
+                # Special-case needed for booleans to turn them lowercase (true/false)
+                if isinstance(attr, bool):
```

Comment:
Since this is Python, we probably want to be even more explicit about limiting the supported types.  For example, if the user sets one of these to a list, we'll get some Python-specific string representation of that out, which is probably not want we want.

We should *also* probably do that type checking in `__init__` so they are caught early.  And I say *also* because that still leaves open the opportunity to change the attributes after creation time, and those wouldn't be typechecked so we need to check them again here in `to_ffi_extra`.

Python dataclasses are sort of designed to address these problems, but they aren't available to us because they were introduced in Python 3.7 and we still support Python 3.6.

Reply:
Question is: what to do with wrong data types? Throw an exception?

Reply:
I put this in an extra commit: [`5357018` (#1603)](https://github.com/mozilla/glean/pull/1603/commits/53570180c4882624c665802b54b51d3497adcd03)
I would probably squash that with the earlier commit, but for easier review iteration it's kept separate for now.